xla/service/collective_pipeliner_test.cc

```diff
@@ -181,6 +181,59 @@ ENTRY entry {
   EXPECT_EQ(get_tuple_index->tuple_index(), 3);
 }
 
+// A case where Bitcast will become the user of a pipelined instruction and
+// check if the DUS is pushed to the next iteration successfully. Absense of
+// Bitcast in acceptable users will break this test.
+TEST_F(CollectivePipelinerTest, BitcastAsUser) {
+  constexpr absl::string_view hlo_string = R"(
+HloModule module
+
+add {
+  lhs = bf16[] parameter(0)
+  rhs = bf16[] parameter(1)
+  ROOT add = bf16[] add(lhs, rhs)
+}
+
+while_cond {
+  param = (s32[], bf16[3,8,128], bf16[3,8,128]) parameter(0)
+  gte = s32[] get-tuple-element(param), index=0
+  constant.1 = s32[] constant(3)
+  ROOT cmp = pred[] compare(gte, constant.1), direction=LT
+}
+
+while_body {
+  param = (s32[], bf16[3,8,128], bf16[3,8,128]) parameter(0)
+  current-loop-index = s32[] get-tuple-element(param), index=0
+  parameter.0 = bf16[3,8,128] get-tuple-element(param), index=1
+  parameter.1 = bf16[3,8,128] get-tuple-element(param), index=2
+  constant.1 = s32[] constant(1)
+  next-loop-index = s32[] add(current-loop-index, constant.1)
+  constant.0 = s32[] constant(0)
+  sliced-parameter.1 = bf16[1,8,128] dynamic-slice(parameter.1, current-loop-index, constant.0, constant.0), dynamic_slice_sizes={1,8,128}
+
+  all-reduce.0 = bf16[1,8,128] all-reduce(sliced-parameter.1), replica_groups={}, to_apply=add, channel_id=1
+  bitcasted-all-reduce.0 = u16[3,8,128] bitcast(all-reduce.0)
+  all-reduce.1 = bf16[3,8,128] bitcast(bitcasted-all-reduce.0)
```

Comment:
This is the one that depends on a bitcast and you want to see it pipelined, right? 
To test your change, I think you need no more than these two ops in the loop, do you?

Reply:
yes this is the part we want to test. but in order to trigger collective pipeliner, we need the pattern of `compute op` (this is what to overlap, dynamic-slice here) `-> collective op` (the other key part to overlap, all-reduce here) `-> elementwise formatting op` (what we are updating, bitcast here) `-> dynamic-update-slice` (because collective pipeliner will reuse the output buffer for intermediate result), I believe it's already the minimal set of ops for this pattern.